source/Halibut.Tests/TcpClientCloseImmediatelyFixture.cs

```diff
@@ -0,0 +1,189 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.Tests.Support;
+using Halibut.Tests.Transport.Streams;
+using Halibut.Tests.Util;
+using Halibut.Transport;
+using NUnit.Framework;
+
+namespace Halibut.Tests
+{
+    public class TcpClientCloseImmediatelyFixture : BaseTest
+    {
+        [Test]
+        public async Task DoesNotWait()
+        {
+            using var writeTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(120));
+            var (disposables, sut, client) = await BuildTcpClientAndTcpListener(
+                CancellationToken,
+                onListenerRead: async _ => await DelayForeverToTryAndDelayWriting(CancellationToken));
+
+            using (disposables)
+            {
+                // Ensure the timeouts are not used
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+
+                var data = new byte[655360];
+                var r = new Random();
+                r.NextBytes(data);
+
+                var semaphore = new SemaphoreSlim(1, 1);
+
+                // Brute force attempt to get the Write to be slow
+                Logger.Information("Start writing");
+                _ = Task.Run(async () =>
+                {
+                    await Try.RunTillExceptionOrCancellation(
+                        async () =>
+                        {
+                            semaphore.Release();
+                            await semaphore.WaitAsync(writeTokenSource.Token);
+                            Logger.Information("WRITER: Start");
+                            await sut.WriteToStream(StreamWriteMethod.WriteAsync, data, 0, data.Length, writeTokenSource.Token);
+                            Logger.Information("WRITER: Finished");
+                        },
+                        CancellationToken);
+                });
+
+                Logger.Information("Wait for writer to block...");
+                // Wait for some brute force writing...
+                while (semaphore.Wait(1000))
```

Comment:
I'm not sure I understand the intent of the waiting logic on `semephore`. 

The intent is to only perform the `lient.CloseImmediately();` once the `await sut.WriteToStream(StreamWriteMethod.WriteAsync, data, 0, data.Length, writeTokenSource.Token);` is blocked and has not returned.

Does this achieve it?

Reply:
It does, but upon review there is still some non-determinism at play here and it's not an intuitive implementation so I'm going to re-implement.